speccurses.py

```diff
@@ -24,90 +26,146 @@
+    :param validator: A validator to ensure only proper keys are entered.
+    :return: A validated value
+    """
     k = 0
     value = list(str(value))
     idx = len(value)
     height, width = stdscr.getmaxyx()
-    pad = curses.newpad(1, width - xoffset)
+    xmax = width * 2
+    pad = curses.newpad(1, xmax)
+    cursorx =  min(x + len(value), width - 1)
 
     while (k != ENTER):
         pad.clear()
-        if(k == BACK and idx > 0):
-            if(idx < len(value)):
-                value = value[:idx-1] + value[idx:]
+        if (k == BACK and idx > 0):
+            if (idx < len(value)):
+                value = value[:idx - 1] + value[idx:]
             else:
                 value = value[:idx - 1]
             idx -= 1
-        elif(k >= 20 and k <= 126 and validator(k)):
-            if(idx == len(value)):
+            if(cursorx > x):
+                cursorx -= 1
+        elif (k >= 20 and k <= 126 and validator(k)):
+            if (idx == len(value)):
                 value += chr(k)
+                if(idx >= xmax):
+                    xmax += width
+                    pad = curses.newpad(1, xmax)
             else:
                 value.insert(idx, chr(k))
             idx += 1
-        elif(k == DEL and idx < len(value)):
-            if(idx > 1):
+            if (cursorx < (width - 1)):
+                cursorx += 1
+        elif (k == DEL and idx < len(value)):
+            if (idx > 1):
                 value = value[:idx] + value[idx + 1:]
             else:
                 value = value[1:]
-        elif(k == curses.KEY_LEFT and idx > 0 ):
+        elif (k == curses.KEY_LEFT and idx > 0):
             idx -= 1
-        elif (k == curses.KEY_RIGHT and idx < len(value) ):
+            if (cursorx > x):
+                cursorx -= 1
+        elif (k == curses.KEY_RIGHT and idx < len(value)):
             idx += 1
-
-
-        stdscr.move(y, idx)
-        pad.addstr(0,0, "".join(value))
-        stdscr.move(y, x + idx)
-        pad.refresh(0,0, y,x, y, width - x)
+            if (cursorx < (width - 1)):
+                cursorx += 1
+        elif(k == curses.KEY_HOME and idx > 0):
+            idx = 0
+            cursorx = x
+        elif (k == curses.KEY_END and idx < len(value)):
+            idx = len(value)
+            cursorx = min(len(value) + x, width - 1)
+
+        stdscr.move(y, cursorx)
+        pad.addstr(0, 0, "".join(value))
+        if(idx != (cursorx - x)):
+            #[0 1 2 3 4 5 6 7]
+            #   o   W
+            #   o X
+            #               I
+            # I - ((W - o) - (W - X))
+            pad.refresh(0, idx - ((width - 1 - x) - (width - 1 - cursorx)), y, x, y, width - 1)
+        else:
+            pad.refresh(0, 0, y, x, y, width - 1)
         k = stdscr.getch()
     return str("".join(value))
 
-#Saves a config file and displays whether or not it was successful
+
+# Saves a config file and displays whether or not it was successful
```

Comment:
This should be formatted as a [docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).